app/scripts/controllers/permissions/methodMiddleware.js

```diff
@@ -0,0 +1,328 @@
+        let accounts = await getAccounts()
+        if (accounts.length > 0) {
+          res.result = accounts
+          return
+        }
+
+        // if no accounts, request the accounts permission
+        try {
+          await requestPermissions({ eth_accounts: {} })
+        } catch (err) {
+          res.error = err
+          return
+        }
+
+        // get the accounts again
+        accounts = await getAccounts()
+        if (accounts.length > 0) {
+          res.result = accounts
+        } else {
+          // this should never happen
+          res.error = ethErrors.rpc.internal(
+            'Accounts unexpectedly unavailable. Please report this bug.'
+          )
+        }
+        return
+
+      // not handled here, just preprocessed to support some syntactic sugar,
+      // specifically 'wallet_plugin'
+      case 'wallet_requestPermissions':
+
+        if (!Array.isArray(req.params)) {
+          res.error = ethErrors.rpc.invalidParams({ data: req })
+          return
+        }
+
+        try {
+          req.params[0] = await preprocessRequestPermissions(req.params[0])
+          break
+        } catch (err) {
+          res.error = err
+          return
+        }
+
+        // plugin-related methods
+
+      // completely handled here
+      case 'wallet_installPlugins':
+
+        if (!Array.isArray(req.params)) {
+          res.error = ethErrors.rpc.invalidParams({ data: req })
+          return
+        }
+
+        try {
+          res.result = await handleInstallPlugins(req.origin, req.params[0])
+        } catch (err) {
+          res.error = err
+        }
+        return
+
+      // basically syntactic sugar for calling a plugin RPC method
+      // we preprocess and forward the request for completion elsewhere
+      case 'wallet_invokePlugin':
+
+        if (
+          !Array.isArray(req.params) ||
+          typeof req.params[0] !== 'string' ||
+          typeof req.params[1] !== 'object' ||
+          Array.isArray(req.params[1])
+        ) {
+          res.error = ethErrors.rpc.invalidParams({ data: req })
+          return
+        }
+
+        req.method = PLUGIN_PREFIX + req.params[0]
+        req.params = [ req.params[1] ]
+        break
+
+      case 'wallet_getPlugins':
+
+        // TODO
+        return
+
+      // a convenience method combining:
+      // - wallet_requestPermissions
+      // - wallet_installPlugins
+      // - eth_accounts (effectively, by parsing returned permissions)
+      case 'wallet_enable':
+
+        if (!Array.isArray(req.params)) {
+          res.error = ethErrors.rpc.invalidParams({ data: req })
+          return
+        }
+
+        const result = {
+          accounts: [],
+          permissions: [],
+          plugins: {},
+        }
+
+        // request the permissions
+
+        try {
+          // we expect the params to be the same as wallet_requestPermissions
+          result.permissions = await requestPermissions(
```

Comment:
Naturally this is also another place where batching would be very valuable, as the popup closing/opening is painfully slow. Not a blocker, just a note (in case it works cheaply).

Reply:
Hmm, I see the point, but in this particular case the permissions request must be fully resolved before we can continue, _unless_ we're saying that a dapp can (1) request a plugin, (2) be denied access to the plugin, and (3) still cause the plugin to be installed.